kep/61-real-load-aware-scheduling/README.md

```diff
@@ -0,0 +1,572 @@
+- Safe Balancing Plugin: It is a node sorting plugin that sorts nodes base on both the mean and the standard deviation of node resource utilization. It aims to balance not only the average load but also the risk caused by load variations.
+
+
+### BestFitBinPack Plugin
+
+
+#### Score plugin
+
+This plugin would extend the Score extension point. K8s scheduler framework calls the Score function for each node separately when scheduling a pod.
+
+Following is the algorithm:
+
+**Algorithm**
+
+1. Get the utilization of the current node to be scored. Call it A.
+2. Calculate the current pod's total CPU requests and overhead. Call it B.
+3. Calculate the expected utilization if the pod is scheduled under this node by adding i.e. U = A + B.
+4. If U &lt;= X%, return U+X as the score
+5. If X% &lt; U &lt;= 100%, return 100 - U
+6. If U > 100%, return 0
+
+For example, let’s say we have three nodes X, Y, and Z, with four cores each and utilization 1, 2, and 3 cores respectively. For simplicity, let’s assume our pod to be scheduled has 0 cores CPU requests and overhead. Let X = 50%.
+
+Utilization of each node:
+
+
+```
+Ux → (1/4)*100 = 25
+Uy → (2/4)*100 = 50
+Uz → (3/4)*100 = 75
+```
+
+
+The score of each node :
+
+
+```
+X → 50 + 25 =  75
+Y → 50 + 50 = 100
+Z → 75 - 50 = 25
+```
+
+
+In the algorithm above, 50% is the target utilization rate we want to achieve on all nodes. We can be less aggressive by reducing it to 40% so that it has much lesser chances of going over 50% during spikes or unexpected loads. 
+So in general to achieve X% utilisation, X - 10 value of U is recommended.
+
+In the 2nd step of the algorithm, one variant uses the current pod's total CPU limits instead of requests, to have a stronger upper bound of expected utilization.
+
+The X% threshold value for utilisation will be made configurable via plugin argument.
+
+**Algorithm Analysis**
+
+<img src="images/bfbp-graph.png" alt="bfbp-graph" width="600" height="400"/>
+
+
+The above is a plot of the piecewise function outlined in the algorithm. The key observations here are manifold:
+
+
+
+1. As the utilization goes from 0 to 50%, we pack pods on those nodes by favoring them.
+2. The nodes are penalized linearly as the utilization goes beyond 50%, to spread the pod amongst those "hot" nodes.
+3. The positive slope begins from 50 and not from 0 because the range of score is from 0-100, and we want to maximize our score output for nodes we wish to favor so that the score is more substantial and other plugins do not affect it much.
+4. There is a break in the graph with a high drop due to the "penalty" we have.
+
+
+### Safe Balancing Plugin
+
+
+#### Score plugin
+
+This plugin would extend the Score extension point. K8s scheduler framework calls the Score function for each node separately when scheduling a pod.
```

Comment:
Is this plugin the same as NodeResourcesBalancedAllocation but with live data?

Reply:
@wangchen615 ^^

Reply:
> Is this plugin the same as NodeResourcesBalancedAllocation but with live data?

@alculquicondor, No. This scoring plugin is taking the usage variations + the usage mean to sort nodes. It's more than balancing the allocation as it also balances the risk of under-provisioning.